apex/contrib/csrc/optimizers/multi_tensor_distopt_adam_kernel.cu

```diff
@@ -194,6 +194,170 @@ struct DistAdamFunctor
+        float update = (next_m_unbiased / denom) + (weight_decay * p[ii]);
+        m[ii] = next_m;
+        v[ii] = next_v;
+        p[ii] -= lr * update;
+      }
+    }
+  }
+
+  __device__ __forceinline__ void operator()(
+    int chunk_size,
+    volatile int* noop_gmem,
+    TensorListMetadata<5>& tl,
+    const float* grad_scale_ptr,
+    const float beta1,
+    const float beta2,
+    const int* step,
+    const int bias_correction,
+    const float eps,
+    const float* lr,
+    adamMode_t mode,
+    const float weight_decay) const
+  {
+    if(*noop_gmem == 1)
+      return;
+
+    float beta1_correction = 1.0f, beta2_correction = 1.0f;
+    if (bias_correction == 1) {
+      beta1_correction = 1 - pow(beta1, *step);
+      beta2_correction = 1 - pow(beta2, *step);
+    }
+
+    int tensor_loc = tl.block_to_tensor[blockIdx.x];
+    int chunk_idx = tl.block_to_chunk[blockIdx.x];
+    int n = tl.sizes[tensor_loc];
+
+    const float grad_scale = *grad_scale_ptr;
+
+    T* p_in = (T *)tl.addresses[0][tensor_loc];
+    p_in += chunk_idx*chunk_size;
+    T* m = (T *)tl.addresses[1][tensor_loc];
+    m += chunk_idx*chunk_size;
+    T* v = (T *)tl.addresses[2][tensor_loc];
+    v += chunk_idx*chunk_size;
+    const GRAD_T* g = (GRAD_T *)tl.addresses[3][tensor_loc];
+    g += chunk_idx*chunk_size;
+    PARAM_OUT_T* p_out = (PARAM_OUT_T *)tl.addresses[4][tensor_loc];
+    p_out += chunk_idx*chunk_size;
+
+    n -= chunk_idx*chunk_size;
+    n = chunk_size < n ? chunk_size : n;
+
+    const bool aligned = (n % ILP == 0 &&
+                          is_aligned(p_in) &&
+                          is_aligned(m) &&
+                          is_aligned(v) &&
+                          is_aligned(g) &&
+                          is_aligned(p_out));
+
+    for (int i_start = threadIdx.x*ILP; i_start < n; i_start += blockDim.x*ILP) {
+      T local_p[ILP];
+      T local_m[ILP];
+      T local_v[ILP];
+      GRAD_T local_g[ILP];
+      PARAM_OUT_T local_p_out[ILP];
+
+      // Load
+      if (aligned) {
+        load_store(local_p, p_in + i_start);
+        load_store(local_m, m + i_start);
+        load_store(local_v, v + i_start);
+        load_store(local_g, g + i_start);
+      } else {
+#pragma unroll
+        for (int ii = 0, i = i_start; ii < ILP; ii++, i++) {
+          if (i < n) {
+            local_p[ii] = p_in[i];
+            local_m[ii] = m[i];
+            local_v[ii] = v[i];
+            local_g[ii] = g[i];
+          } else {
+            local_p[ii] = 0;
+            local_m[ii] = 0;
+            local_v[ii] = 0;
+            local_g[ii] = 0;
+          }
+        }
+      }
+
+      // Local compute
+      local_step(
+        local_p, local_m, local_v, local_g, grad_scale,
+        beta1, beta2, beta1_correction, beta2_correction,
+        eps, *lr, mode, weight_decay);
+#pragma unroll
+      for (int ii = 0; ii < ILP; ii++) {
+        local_p_out[ii] = static_cast<PARAM_OUT_T>(local_p[ii]);
+      }
+
+      // Store
```

Comment:
would there be any appetite to use gradients after `step`? if so, it'd be necessary to store unscaled gradients as well.

Reply:
We haven't encounter such cases yet, after optimizer stepping a new iteration starts, where the gradients will be zero out first. This is the same behavior as other optimizer kernels in the repo, so I think we can leave it as it is until there're cases we need to store unscaled gradients in future.